pkg/cluster/deploybaseresources.go

```diff
@@ -254,65 +256,84 @@ func (m *manager) subnetsWithServiceEndpoint(ctx context.Context, serviceEndpoin
 	return subnets, nil
 }
 
-func (m *manager) attachNSGs(ctx context.Context) (bool, error) {
+func (m *manager) attachNSGs(ctx context.Context) error {
+	return m._attachNSGs(ctx, 3*time.Minute)
+}
+
+func (m *manager) _attachNSGs(ctx context.Context, timeout time.Duration) error {
 	if m.doc.OpenShiftCluster.Properties.NetworkProfile.PreconfiguredNSG == api.PreconfiguredNSGEnabled {
-		return true, nil
+		return nil
 	}
+	var innerErr error
+
 	workerProfiles, _ := api.GetEnrichedWorkerProfiles(m.doc.OpenShiftCluster.Properties)
 	workerSubnetId := workerProfiles[0].SubnetID
 
-	for _, subnetID := range []string{
-		m.doc.OpenShiftCluster.Properties.MasterProfile.SubnetID,
-		workerSubnetId,
-	} {
-		m.log.Printf("attaching network security group to subnet %s", subnetID)
-
-		// TODO: there is probably an undesirable race condition here - check if etags can help.
-
-		s, err := m.subnet.Get(ctx, subnetID)
-		if err != nil {
-			return false, err
-		}
+	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	_ = wait.PollImmediateUntil(30*time.Second, func() (bool, error) {
```

Comment:
It would be good to document the significance of 30 seconds with a comment. I don't think this should go inside the for loop - there's only two subnets being ranged over and the point seems to be that there may be errors at different stages of the process that require retrying from the top.